### PR TITLE
ColumnCollection Refactor, better locking around shared Tx

### PR DESCRIPTION
**Pull request by wcharczuk**
This PR integrates some changes I made around the column collection, adding more efficient ways to read data from queries by giving joins their own prefix, and adds a bunch of locking around the shared connection transaction.

### CODE REVIEW

File: spiffy.go [hunk 1]
```diff
@@ -1031,12 +1116,21 @@ func (dbAlias *DbConnection) GetAllInTransaction(collection interface{}, tx *sql
...
+	var popErr error
 	for rows.Next() {
 		newObj, _ := MakeNew(t)
-		popErr := PopulateInOrder(newObj, rows, meta)
-		if popErr != nil {
-			err = exception.Wrap(popErr)
-			return
+
+		if isPopulatable {
+			popErr = (AsPopulatable(newObj)).Populate(rows)
```
Comment (John-Steidley) [inline spiffy.go]: Remove extra parens

File: spiffy.go [hunk 2]
```diff
@@ -1284,10 +1407,100 @@ func (dbAlias *DbConnection) DeleteInTransaction(object DatabaseMapped, tx *sql.
...
+
+	dbAlias.TxLock.Lock()
+	defer dbAlias.TxLock.Unlock()
+	dbAlias.Tx = tx
+}
+
+// ReleaseIsolation reverses `IsolateToTransaction`
+func (dbAlias *DbConnection) ReleaseIsolation() {
+	if dbAlias == nil {
+		panic("`dbAlias` is nil; did you remember to set up the DbConnection in your project?")
```
Comment (John-Steidley) [inline spiffy.go]: This should be a DBAliasNilError

File: spiffy.go [hunk 3]
```diff
@@ -523,10 +604,14 @@ func (q *QueryResult) Out(object DatabaseMapped) (err error) {
 		return
 	}
 
-	meta := NewColumnCollectionFromInstance(object)
-
+	columnMeta := NewColumnCollectionFromInstance(object)
+	var popErr error
 	if q.Rows.Next() {
-		popErr := PopulateByName(object, q.Rows, meta)
+		if populatable, isPopulatable := object.(Populatable); isPopulatable {
```
Comment (John-Steidley) [inline spiffy.go]: This block is different from the other ones. Let's use the new utility functions here as well.

**Comment by John-Steidley** [comment]
I think it's time we split spiffy into separate files. There seems to be 3-4 divisions based on the source code. What do you think about Column, ColumnCollection, QueryResult, and DbConnection all living in separate files?

**Comment by wcharczuk** [comment]
I wish there was threading in Github comments.

Yeah, I think it's gotten to the point where `spiffy.go` is sufficiently unwieldy that it'll be good to split it up. The big lift is matching up the tests but I'll try and crank on that this afternoon.

File: spiffy.go (hunk 3 above)
Comment (wcharczuk) [inline spiffy.go]: so we actually want to keep this because it prevents doing the type assertion twice.

**Comment by wcharczuk** [comment]
Files are split

**Comment by wcharczuk** [comment]
@John-Steidley i'd like to merge this at some point tomorrow if possible

File: column.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+						}
+					} else {
+						convertedValue := valueReflected.Convert(fieldType)
+						field.Set(convertedValue)
+					}
+				}
+			}
+		}
+	} else {
+		return exception.New("hit a field we can't set: '" + c.FieldName + "', did you forget to pass the object as a reference?")
```
Comment (John-Steidley) [inline column.go]: Handle this error up front.

File: spiffy.go (hunk 3 above)
Comment (John-Steidley) [inline spiffy.go]: Good point. In that case db_connection.go should do the same thing in GetByIDInTransaction.

Which in turn makes me suspicious of the value of AsPopulatable and IsPopulatable. I think the casts are more clear.

File: db_connection.go [hunk 5]
```diff
@@ -0,0 +1,792 @@
...
+// --------------------------------------------------------------------------------
+
+// NewUnauthenticatedDbConnection creates a new DbConnection without Username or Password or SSLMode
+func NewUnauthenticatedDbConnection(host, schema string) *DbConnection {
+	conn := &DbConnection{}
+	conn.Host = host
+	conn.Schema = schema
+	conn.Username = ""
+	conn.Password = ""
+	conn.SSLMode = "disable"
```
Comment (John-Steidley) [inline db_connection.go]: We should write these to call each other. I don't like that we had to add the new TxLock in four places.

File: db_connection.go [hunk 6]
```diff
@@ -0,0 +1,792 @@
...
+			recoveryException := exception.New(r)
+			err = exception.WrapMany(err, recoveryException)
+		}
+	}()
+
+	if dbc == nil {
+		return exception.New(DBAliasNilError)
+	}
+
+	dbc.txLock()
```
Comment (John-Steidley) [inline db_connection.go]: Why is this here?

Comment (John-Steidley) [inline db_connection.go]: Put differently: what's the purpose of the TxLock generally? It seems to be trying to protect the Tx field on a DbConnection, but despite being RWLock, it isn't aquired for all reads. It's overall a very complicated structure. Could you try to explain it further so that I can try to audit its behavior?

Comment (wcharczuk) [inline db_connection.go]: The purpose is to serialize access to the transaction tied to a connection.

It is a RW lock to facilitate checks on `IsIsolatedToTransaction` vs. almost every other case where we're writing to the transaction metadata effectively. 

We use this structure because we only conditionally need to synchronize (really only during controller tests). If the connection is not isolated, none of these locks take effect (why we have separate txLock and txUnlock methods that check the tx). 

I agree it's hugely complicated, but short of being able to pass transactions in testing scenarios to controllers, we need to have some kind of shared state. That shared state needs to be synchronized.

Comment (John-Steidley) [inline db_connection.go]: It seems to me that there are 4 places where ReleaseIsolation can race with an unsynchronized access to dbc.Tx (Prepare, Begin, txLock, and txUnlock). Shouldn't we wrap each of those accesses with RLock and RUnlock pairs? If not, why not?

File: query_result.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+	"github.com/blendlabs/go-exception"
+)
+
+// --------------------------------------------------------------------------------
+// Query Result
+// --------------------------------------------------------------------------------
+
+// QueryResult is the intermediate result of a query.
+type QueryResult struct {
```
Comment (John-Steidley) [inline query_result.go]: QueryResult is unnecessarily public, so are its fields.

Comment (wcharczuk) [inline query_result.go]: QueryResult was public initially because I wanted to be able to save references to it for intermediate steps, making the fields private is the right move though.

File: db_connection.go (hunk 6 above)
Comment (wcharczuk) [inline db_connection.go]: In prepare we don't want to acquire locks because the responsibility of unlocking is unclearly assigned. The caller of prepare should be handling the locks. 

For Begin, that is an error and has been corrected (`dbc.Tx != nil` => `dbc.IsIsolatedToTransaction()`)

txLock and txUnlock are called so often that I don't want to put a lock acq in the normal path of when we don't use isolation (i.e. normal run mode vs. testing).

**Comment by John-Steidley** [comment]
LGTM
